nmdc_runtime/site/resources.py

```diff
@@ -371,30 +371,47 @@ def gold_api_client_resource(context: InitResourceContext):
 
 @dataclass
 class NmdcPortalApiClient:
+
     base_url: str
-    # Using a cookie for authentication is not ideal and should be replaced
-    # when this API has an another authentication method
-    session_cookie: str
+    refresh_token: str
+    access_token: Optional[str] = None
+    access_token_expires_at: Optional[datetime] = None
+
+    def _request(self, method: str, endpoint: str, **kwargs):
+        if self.access_token is None or datetime.now() > self.access_token_expires_at:
+            refresh_response = requests.post(
+                f"{self.base_url}/auth/refresh",
+                json={"refresh_token": self.refresh_token},
+            )
+            refresh_response.raise_for_status()
+            refresh_body = refresh_response.json()
+            self.access_token_expires_at = datetime.now() + timedelta(
+                seconds=refresh_body["expires"]
```

Comment:
Side note (not the focus of this PR): I'm surprised that the Data Portal API returns the token's expiration as an offset, instead of as an absolute time. It seems to me that that could lead to there being an offset between the time the client thinks it expires and the time the server thinks it expires (e.g. a client on a slow internet connection might think it expires a couple seconds later than the server does).

Reply:
`nmdc-server` is _not_ a full OAuth 2.0 authorization server, but the `expires` field is inline with OAuth 2.0 recommendation of sending the lifetime in second: https://datatracker.ietf.org/doc/html/rfc6749#section-5.1. I don't know if using `expires` instead of `expires_in` was intentional or an oversight.

Reply:
I think it was an oversight so I'm requesting we change it to `expires_in`: https://github.com/microbiomedata/nmdc-server/pull/1287 Once that's accepted and merged I'll updated this PR as well.

Reply:
Thanks for providing links to the recommendation. I agree that calling it `expires_in` (which is what it's called in the latest commit on this branch) and having it be a number of seconds (as opposed to an absolute timestamp) is in line with the recommendation.